win32ss/user/ntuser/ghost.c

```diff
@@ -1,18 +1,195 @@
 /*
  * PROJECT:     ReactOS user32.dll
  * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
- * PURPOSE:     Ghost window handling
+ * PURPOSE:     Window ghosting feature
  * COPYRIGHT:   Copyright 2018 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
  */
 
 #include <win32k.h>
+#include "ghostwnd.h"
 
-BOOL FASTCALL IntGoGhost(PWND Window, BOOL bGo)
+#define NDEBUG
+#include <debug.h>
+
+BOOL FASTCALL LookupFnIdToiCls(int FnId, int *iCls);
+
+INT
+UserGetClassName(IN PCLS Class,
+                 IN OUT PUNICODE_STRING ClassName,
+                 IN RTL_ATOM Atom,
+                 IN BOOL Ansi);
+
+static UNICODE_STRING GhostClass = RTL_CONSTANT_STRING(GHOSTCLASSNAME);
+static UNICODE_STRING GhostProp = RTL_CONSTANT_STRING(GHOST_PROP);
+
+BOOL FASTCALL IntIsGhostWindow(PWND Window)
 {
+    BOOLEAN Ret = FALSE;
+    UNICODE_STRING ClassName;
+    INT iCls, Len;
+    RTL_ATOM Atom = 0;
+
+    if (!Window)
+        return FALSE;
+
+    if (Window->fnid && !(Window->fnid & FNID_DESTROY))
+    {
+        if (LookupFnIdToiCls(Window->fnid, &iCls))
+        {
+            Atom = gpsi->atomSysClass[iCls];
+        }
+    }
+
+    // check class name
+    _SEH2_TRY
+    {
+        RtlInitUnicodeString(&ClassName, NULL);
+        Len = UserGetClassName(Window->pcls, &ClassName, Atom, FALSE);
+        if (Len > 0)
+        {
+            Ret = RtlEqualUnicodeString(&ClassName, &GhostClass, TRUE);
+        }
+        else
+        {
+            DPRINT1("Unable to get class name\n");
+        }
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        DPRINT1("Exception!\n");
+        Ret = FALSE;
+    }
+    _SEH2_END;
+
+    RtlFreeUnicodeString(&ClassName);
+
+    return Ret;
+}
+
+HWND APIENTRY NtUserGhostWindowFromHungWindow(HWND hwndHung)
+{
+    RTL_ATOM Atom;
+    PWND pHungWnd;
+    PPROPERTY Prop;
+    HWND hwndGhost;
+
+    pHungWnd = ValidateHwndNoErr(hwndHung);
+    if (!pHungWnd)
+    {
+        DPRINT("Not a window\n");
+        return NULL;
+    }
+
+    if (!IntGetAtomFromStringOrAtom(&GhostProp, &Atom))
+        ASSERT(FALSE);
+
+    Prop = UserGetProp(pHungWnd, Atom, TRUE);
+
+    _SEH2_TRY
+    {
+        ProbeForRead(Prop, sizeof(PROPERTY), 1);
```

Comment:
Prop is expected to be a kernel pointer, so it shouldn't need SEH, and definitely can't have a Probe

Reply:
OK.